proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,237 @@
+the bank runtime will halt and wait until the computation is finished, before
+processing any blocks with height greater than `N`.
+
+At the end of the reward computation phase, a set of sysvar accounts will be
+updated - `EpochRewardProof` and `RewardReserve`.
+
+`EpochRewardProof` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardProof` is an array of entries,
+```
+[(epoch, (reward_amount, root_hash))]
+```
+, where each entry corresponds to the reward for
+one particular epoch. The maximum depth of the history are the same as the
+depth of stake history, which is 512. The `root_hash` of the merkle tree for
+all rewards are optional. It is for potential future works to distribute
+rewards as transactions.
+
+When reaching the end of reward computation phase, `EpochRewardProof` are
+updated with the reward computation result. A new entry will be pushed into the
+array. If the length of the history array reaches the maximum depth, the oldest
+entry will popped.
+
+```
+struct RewardHistoryEntry {
+   total: u64,                  // total rewards for this epoch
+   distributed: u64,            // already distributed reward amount
+   root_hash: Option<Hash>,     // root-hash of reward merkle tree
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+At the end of reward computation phase, a set of `EpochRewardsReserve` sysvar
+accounts will be updated too. Each `RewardsReserve` records the reward balance
+to be distributed for a particular block in `Reward Distribution` phase, which
+captures a particular subset of stakes rewards to be paid at the `index` block in
+the reward distribution phase. The complete subset of stake rewards forms a
+partition of the all the stake rewards.
+
+```
+struct EpochRewardReserve {
+   balance: u64,
+   index: u64,
+   hash: Hash
+}
+```
+
+`balance` is total amount of rewards (in lamports) to be distributed in this block.
+`index` is the reward distribution block height offset from reward computation end block.
+`hash` is the hash of all distributed rewards for this block.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks. For each block, `i` in reward distribution phase, the
+rewards correspond to `EpochRewardReserve` sysvar account with `index=i` will
+be paid out, approximate `1/M` of the total stake reward accounts.
+
+This partition is required to be deterministic for the current epoch, but also
+random across different epochs. This can be achieved by hashing the account's
+pubkey with some epoch dependent values, sort the results, and divide them into
+`M` bins. The epoch dependent value can be the epoch number, total rewards for
+the epoch, the leader pubkey for the epoch block, etc. `M` are chosen based on
+50K account per block, which equal to `ceil(num_stake_accounts/50,000)`.
```

Comment:
> Why 50k accounts per block? Would it make sense to include that reasoning in this proposal? Do we need to add any levers-to-pull in order to change this amount later (or just deal with it later)?

Good question. 

Looking at the most recent epoch stats on mainnet, it takes about 10 seconds to store 592K stake rewards. 50K reward store will be around 800ms, which is a reasonable overhead for block times during reward distribution phase.

Reply:
800ms more.

Reply:
100%, target block time really should remain 400ms

Reply:
goal: network remains butter across the epoch boundary

Reply:
4000 accounts per block sounds good me. 

Another overhead is that there will be 150 `EpochRewardReserve` accounts, which should be fine too.

Reply:
sure that seems ok. I'm tempted to consider a more dynamic approach to adjust to changing node performance over time, but I feel this could be a follow-up SIMD if needed down the road.

`ACCOUNTS_PER_TICK` might be a nice way to express this number (~64 accounts per tick)